### ISSUE TITLE
Filter page / post lists based on permalink?

### ISSUE THREAD

**Issue by miztroh** [WP-API/WP-API#1776]
Is it possible to query for pages / posts and filter by permalink?  Something along the lines of ``/wp/v2/posts?filter[permalink]=/2015/11/29/hello-world/``?

**Comment by rmccue** [comment]
This isn't possible currently, and I don't think it's possible in the API, as the functionality that turns a permalink into a query is contained in `WP` (the class), which uses the superglobals directly.

One possibility though is that we could send a `Link` header/`<link>` element on post pages, which would link to the REST API endpoint for it; something like:
```
Link: http://example.com/wp-json/wp/v2/posts/4; rel=alternate; type=application/json
```

This doesn't exist right now though.

Can you elaborate on what exactly you're trying to do? It's useful for us to understand the use cases for things like this :)

**Comment by jdigory** [comment]
I wanted to do something similar, but for slug. If I want to mirror the URL structure in my own app, via the API, then I would allow users to hit URLs like `/2015/11/29/hello-world`. The slug in this case is `hello-world`. 

So, how to find the corresponding post? I was previously retrieving all posts, and then locating from the resulting array, but this is cumbersome.

Instead, this API request will filter by slug: `/wp-json/wp/v2/posts?filter[name]=hello-world`

Perhaps this will work for you as well. (Beats me why the filter param is `name` when there is no corresponding key in the json, but it works nonetheless.)

**Comment by danielbachhuber** [comment]
> So, how to find the corresponding post? I was previously retrieving all posts, and then locating from the resulting array, but this is cumbersome.

I think the unfortunate truth about clients right now is that you need to completely replicate WordPress' rewrite rules in whichever routing mechanism you're using. As such, you need to make your app transform `/2015/11/29/hello-world` into `year=2015&monthnum=11&day=29&name=hello-world`, which you can then pass to WP-API through `filter`

**Comment by miztroh** [comment]
@rmccue Basically, I'd love to write a custom elements (specifically Polymer) theme that handles all the URL parsing, routing, page loading on the client side such that all I have have to do on the PHP side is generate a nonce.  Ideally, I'd like to query the API with the current URL and have it return information about the sort of content located at that URL (post, page, 404, search results, etc).

**Comment by yantakus** [comment]
@rmccue I've described my use-case in detail here: https://github.com/WP-API/WP-API/issues/2164

@danielbachhuber 

> As such, you need to make your app transform /2015/11/29/hello-world into year=2015&monthnum=11&day=29&name=hello-world, which you can then pass to WP-API through filter

If WP can do it by extracting from the URL, I think it is easy to create a new API endpoint which will expose this functionality and return the content behind a certain URL. Isn't it?

**Comment by danielbachhuber** [comment]
> I think it is easy to create a new API endpoint which will expose this functionality and return the content behind a certain URL. Isn't it?

It's possible with code, yes. However, it's not the concern of the API to handle routing.

**Comment by miztroh-zz** [comment]
@danielbachhuber I disagree.  WP already has a system in place to determine page information based on its URL (is it a post, is it a page, is it a 404, etc.).  That system should be exposed via the API so that frontend developers using the API don't have to come up with their own logic for determining page information.  Developers should be able to create WP frontends that don't require any server code (other than nonce / auth stuff).  IMO, this is a critical piece.

**Comment by kadamwhite** [comment]
> Developers should be able to create WP frontends that don't require any server code

Technically, this is true now -- if your entire theme is a single index.php which loads a front-end application, that application should then be able to handle all its own routing from that point on.

I can see your argument, and I could imagine ways you could make an endpoint which could map a URL to the WordPress context that would normally be exposed on that page, but I agree with @danielbachhuber and @rmccue that doing so would be a departure from the API's goals. Routing in the traditional WP sense originates from a very different paradigm than a resource-based API.

**Comment by yantakus** [comment]
> if your entire theme is a single index.php which loads a front-end application, that application should then be able to handle all its own routing from that point on.

But in this case it is impossible to create a fully fledged WP SPA theme, because it's routing behavior won't depend on permalinks settings specified in WP admin panel.

Imagine a wordpress developer created a commercial wordpress SPA theme. It should work out of the box and let users modify any content via admin panel without knowledge of JS/CSS/HTML. But it will be impossible if the theme's routing settings don't depend on the settings specified in the admin panel. In this case if a user changes permalinks settings in the dashboard it will require additional configuration inside the theme. Isn't it weird?

The only reasonable way to handle SPA theme's routing I see is to replicate the routes a user sees in WP admin panel when creates a post/page/etc. If he/she follows the link and sees another page, this would be silly.

If you don't think routing is a concern of API, you should give developers another way to avoid the problem described above.

**Comment by yantakus** [comment]
Will this ever be implemented? Can't find anything like this in Trac.

**Comment by adampatterson** [comment]
> I think the unfortunate truth about clients right now is that you need to completely replicate WordPress' rewrite rules in whichever routing mechanism you're using. As such, you need to make your app transform /2015/11/29/hello-world into year=2015&monthnum=11&day=29&name=hello-world, which you can then pass to WP-API through filter

@danielbachhuber It is not always possible to replicate WordPress routing.

Expanding a bit on my comments here https://github.com/WP-API/WP-API/issues/456#issuecomment-360958166

The site that I am consuming by API into Laravel has a structure like this.

* site.com/{category}
* site.com/{tag}
* site.com/{post-name}

*Agreed that's a terrible structure, but it's not my doing. But many sites are older than the WP API Concept.*

The reason why it would be better or at least should be considered to look for content by URI is that I honestly can't return data in any other way. Unless I change the structure of the app I'm working with. Presumably would require me to modify the content returned by the API, also no way of me know what I would be changing from or to.

Meaning any post linking to a category in an external route based system requires that I have to check to see if it's a post/category/tag before I can decide what to do.  Or if I am trying to return subpages I have to look the parent page up, then look up the subpage ( possible returning 2 or more results ) then check to see what matches the parent ID.

In short ( lol ), the WordPress routing AFAIK uses the GUID in the posts table to decide if something is a post, page, or attachment. 

I don't see this as a departure from the API's goals ( not sure what they are cc @kadamwhite ), But I see this as a way to query by an already used field in the WordPress core.
